libs/datamodel/core/src/ast/mod.rs

```diff
@@ -87,15 +94,67 @@ impl SchemaAst {
     }
 }
 
-/// An opaque identifier for a top-level item in a schema AST. Use the
-/// `schema[top_id]` syntax to resolve the id to an `ast::Top`.
+/// An opaque identifier for a model in a schema AST. Use the
+/// `schema[model_id]` syntax to resolve the id to an `ast::Model`.
 #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub(crate) struct TopId(u32);
+pub(crate) struct ModelId(u32);
+
+impl std::ops::Index<ModelId> for SchemaAst {
+    type Output = Model;
+
+    fn index(&self, index: ModelId) -> &Self::Output {
```

Comment:
Can we track caller here so we get better stacktraces?

Reply:
That crossed my mind as well, but I'm not sure how that all combines with the `schema[...]` sugar. I'll do a quick experiment in a dummy crate and report pack.

Reply:
Test program:

```rust
use std::ops::Index;

fn main() {
    let s = MyStruct {
        myfields: &[
            "doc", "grumpy", "sleepy", "happy", "bashful", "sneezy", "dopey",
        ],
    };

    assert_eq!(s.myfields[3], "happy");
    let idx = MyIndex(3);
    assert_eq!(s[idx], "happy");

    s[MyIndex(1000)];
    s.myfields[900];

    println!("Ok")
}

struct MyStruct {
    myfields: &'static [&'static str],
}

struct MyIndex(u32);

impl Index<MyIndex> for MyStruct {
    type Output = &'static str;

    fn index(&self, index: MyIndex) -> &Self::Output {
        &self.myfields[index.0 as usize]
    }
}
```

Result:

```
thread 'main' panicked at 'index out of bounds: the len is 7 but the index is 1000', src/main.rs:14:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

So the `#[track_caller]` isn't necessary there.